src/seraphis_mocks/scan_context_async_mock.cpp

```diff
@@ -0,0 +1,797 @@
+    async::join_token_t data_join_token                             = m_threadpool.get_join_token(data_join_signal);
+
+    auto task =
+        [
+            this,
+            l_chunk_request                = chunk_request,
+            l_context_stop_flag            = context_stop_signal.get_future().share(),
+            l_data_stop_flag               = data_stop_signal.get_future().share(),
+            l_chunk_context                = std::make_shared<std::promise<sp::scanning::ChunkContext>>(std::move(chunk_context_handle)),
+            l_chunk_data                   = std::make_shared<std::promise<sp::scanning::ChunkData>>(std::move(chunk_data_handle)),
+            l_context_join_token           = context_join_token,
+            l_data_join_token              = data_join_token
+        ]() mutable -> async::TaskVariant
+        {
+            return chunk_task(l_chunk_request,
+                l_context_stop_flag,
+                l_data_stop_flag,
+                l_chunk_context,
+                l_chunk_data,
+                l_context_join_token,
+                l_data_join_token);
+        };
+
+    // launch the task
+    m_num_pending_chunks.fetch_add(1, std::memory_order_relaxed);
+    m_threadpool.submit(async::make_simple_task(async::DefaultPriorityLevels::MEDIUM, std::move(task)));
+
+    // return pending chunk for caller to deal with as needed
+    async::join_condition_t chunk_context_join_condition{
+            m_threadpool.get_join_condition(std::move(context_join_signal), std::move(context_join_token))
+        };
+
+    async::join_condition_t chunk_data_join_condition{
+            m_threadpool.get_join_condition(std::move(data_join_signal), std::move(data_join_token))
+        };
+
+    return PendingChunk{
+            .chunk_request = chunk_request,
+            .pending_context = sp::scanning::PendingChunkContext{
+                    .stop_signal            = std::move(context_stop_signal),
+                    .chunk_context          = std::move(chunk_context_future),
+                    .context_join_condition = std::move(chunk_context_join_condition)
+                },
+            .pending_data    = sp::scanning::PendingChunkData{
+                    .stop_signal            = std::move(data_stop_signal),
+                    .chunk_data             = std::move(chunk_data_future),
+                    .data_join_condition    = std::move(chunk_data_join_condition)
+                }
+        };
+}
+//-------------------------------------------------------------------------------------------------------------------
+void AsyncScanContextLegacy::launch_next_chunk_task(const std::unique_lock<std::mutex> &pending_queue_lock)
+{
+    THROW_WALLET_EXCEPTION_IF(!pending_queue_lock.owns_lock(), tools::error::wallet_internal_error,
+        "Pending queue must be locked to launch the next chunk task");
+
```

Comment:
```suggestion
bool AsyncScanContextLegacy::try_launch_next_chunk_task(const std::unique_lock<std::mutex> &pending_queue_lock)
{
    THROW_WALLET_EXCEPTION_IF(!pending_queue_lock.owns_lock(), tools::error::wallet_internal_error,
        "Pending queue must be locked to launch the next chunk task");

    if !check_launch_next_task(lock)
        return false;
```
I'd do it like this. Then when `launch_next_chunk_task` is called below it can instead be called directly.

You can also move this debug statement to the top of this function: `MDEBUG("Attempting to launch chunk task at ...`

Reply:
Also the `pending_queue_lock.owns_lock(), ` check isn't strictly safe, because it could be *any* lock passed in. You probably need a thread-local id + atomic member variable set to equal the thread id that has the lock (with scoped leave handler that unsets the thread id before dropping the top-level lock guard). Then check the locked thread id equals the current thread's id.

Reply:
> because it could be any lock passed in

It's only ever called with a thread-local unique lock. When I see a unique lock, I understand them to be thread-local ([they're supposed to be scoped to a thread](https://stackoverflow.com/a/40665203)).

I'm just going to remove this check instead of implementing the suggestion if this is a no go as is. It's there to provide some clarity on how the scoped lock is expected to be controlled. If it doesn't help provide that clarity, I'm just going to remove it

Reply:
> It's only ever called with a thread-local unique lock. When I see a unique lock, I understand them to be thread-local ([they're supposed to be scoped to a thread](https://stackoverflow.com/a/40665203)).

Right, but there is no guarantee in the check that you locked the right mutex. It's just implied/encouraged by the code structure. Someone reading the code would still need to validate where the lock comes from. So I'd add a comment explaining what the lock is supposed to hold.

Reply:
Implemented [this](https://github.com/UkoeHB/monero/pull/23#discussion_r1578678809) in the latest